### PR TITLE
Add support stateful precompiles

### PR DESCRIPTION
**Pull request by fbac**
Major changes
----
- Cherry-pick the changes that introduced stateful precompiled contracts into go-ethereum upstream.

### CODE REVIEW

File: core/vm/contract.go [hunk 1]
```diff
@@ -82,7 +84,34 @@ func NewContract(caller ContractRef, object ContractRef, value *big.Int, gas uin
...
+	c := &Contract{
+		CallerAddress: caller.Address(),
+		caller:        caller,
+		self:          object,
+		isPrecompile:  true,
+	}
+
+	// Gas should be a pointer so it can safely be reduced through the run
+	// This pointer will be off the state transition
+	c.Gas = gas
```
Comment (skosito) [inline core/vm/contract.go:98]: should we init jumpdests as well as above constructor? if parent doesnt make sense maybe simple map init?

or jumpdest is completely disregarded in this case?

File: eth/tracers/js/goja.go [hunk 2]
```diff
@@ -238,7 +238,7 @@ func (t *jsTracer) CaptureStart(env *vm.EVM, from common.Address, to common.Addr
 	t.ctx["block"] = t.vm.ToValue(env.Context.BlockNumber.Uint64())
 	// Update list of precompiles based on current block
 	rules := env.ChainConfig().Rules(env.Context.BlockNumber, env.Context.Random != nil)
-	t.activePrecompiles = vm.ActivePrecompiles(rules)
+	t.activePrecompiles = env.ActivePrecompiles(rules)
```
Comment (skosito) [inline eth/tracers/js/goja.go:241]: what is the difference between changing to DefaultActivePrecompiles change above in runtime.go, and using ActivePrecompiles here?

File: internal/ethapi/api.go [hunk 3]
```diff
@@ -1410,7 +1410,7 @@ func AccessList(ctx context.Context, b Backend, blockNrOrHash rpc.BlockNumberOrH
 	}
 	isPostMerge := header.Difficulty.Cmp(common.Big0) == 0
 	// Retrieve the precompiles since they don't need to be added to the access list
-	precompiles := vm.ActivePrecompiles(b.ChainConfig().Rules(header.Number, isPostMerge))
+	precompiles := vm.DefaultActivePrecompiles(b.ChainConfig().Rules(header.Number, isPostMerge))
```
Comment (skosito) [inline internal/ethapi/api.go:1413]: same question about DefaultActivePrecompiles?

**Review by skosito — COMMENTED** [review]
just small comments, otherwise looks good but unclear how to test it, CI doesn't contain tests?

**Comment by fbac** [comment]
> just small comments, otherwise looks good but unclear how to test it, CI doesn't contain tests?

The CI doesn't contain any workflow currently. Tests in localhost are successful except for the DNS discovery package, which requires setup. I'll be opening issues to set up CI and also Makefile targets to set up our envs.

File: internal/ethapi/api.go (hunk 3 above)
Comment (fbac) [reply internal/ethapi/api.go:1413]: This is just a rename from ActivePrecompiles to DefaultActivePrecompiles, but it's still the same func.

File: core/vm/contract.go (hunk 1 above)
Comment (fbac) [reply core/vm/contract.go:98]: Stateful precompiled contracts are in the end golang code executed when someone calls the blockchain at an specific address. As we're not going to execute JUMP through the EVM, we shouldn't need to initialize it.

In fact take a look at this bit:
```
func (c *Contract) validJumpdest(dest *uint256.Int) bool {
	if c.isPrecompile {
		return false
	}
       ...
}
```

File: internal/ethapi/api.go (hunk 3 above)
Comment (skosito) [reply internal/ethapi/api.go:1413]: @fbac  it its rename, is this ok, please double check before merge:

```
t.activePrecompiles = env.ActivePrecompiles(rules)
```

**Review by skosito — APPROVED** [review]

File: internal/ethapi/api.go (hunk 3 above)
Comment (fbac) [reply internal/ethapi/api.go:1413]: Yes, it's correct. In this case what that does is returning the EVM precompiles field (evm.activePrecompiles), which in turn is populated with DefaultActivePrecompiles:
```
evm.activePrecompiles = DefaultActivePrecompiles(evm.chainRules)
```

`ActivePrecompiles` is an accessor method to evm.activePrecompiles.
`DefaultActivePrecompiles` is a func that returns the set of active precompiled contracts for a fork (berlin, etc) and is used to populate evm.activePrecompiles.

**Pull request merged**